### DISCUSSION TITLE
BGZF compression support for CSV format

### DISCUSSION THREAD

**Discussion by antonylebechec** [duckdb/duckdb#6716] [Ideas]
Hi DuckDB team!

Working in bioinformatics, I deal with huge amount of data using BGZF compression format (or BGZIP). BGZF is block compression implemented on top of the standard gzip file format.

This BGZF format is especially used to compress VCF files (genomic variants, a CSV tab-delimited file with header). As uncompressed VCF files are perfectly read in duckdb, as well as GZIP VCF files (never used), BGZF are not (yet?) supported.

Some solution are used to deal with these huge amount of data (tens or hundreds of Gb), such as re-compression (gzip), Python module (Bio.bgzf and pandas or dask), these steps take a while...

It would be very useful if BGZF format was taken into account! It is close to GZIP, but I can't assess if it's a great effort.

To help, if you decide to implement BGZF format, some files attached.

Let me know!

Best,

--
References:
BGZF in Python: https://biopython.org/docs/1.75/api/Bio.bgzf.html
BGZF in samtools: http://samtools.github.io/hts-specs/SAMv1.pdf

Code:
```
import duckdb
file_csv = "tests/data/example.vcf.csv"
file_gzip = "tests/data/example.vcf.gzip.csv.gz"
file_bgzip = "tests/data/example.vcf.bgzip.csv.gz"
con = duckdb.connect(':memory:')
con.execute(f""" CREATE TABLE csv AS SELECT * FROM '{file_csv}' """)
con.execute(f""" CREATE TABLE csv_gzip AS SELECT * FROM '{file_gzip}' """)
con.execute(f""" CREATE TABLE csv_bgzip AS SELECT * FROM '{file_bgzip}' """)
```

Error:
```
Traceback (most recent call last):
  File "/mnt/BIOINFO/git/HOWARD/tests/test.external_db.py", line 36, in <module>
    con.execute(f""" CREATE TABLE csv_bgzip AS SELECT * FROM '{file_bgzip}' """)
duckdb.IOException: IO Error: Unsupported GZIP archive
```

Files:
[example.vcf.bgzip.csv.gz](https://github.com/duckdb/duckdb/files/10971700/example.vcf.bgzip.csv.gz)
[example.vcf.csv](https://github.com/duckdb/duckdb/files/10971701/example.vcf.csv)
[example.vcf.gzip.csv.gz](https://github.com/duckdb/duckdb/files/10971702/example.vcf.gzip.csv.gz)

**Comment by hannes** [comment]
Does not look to be that crazy actually, just some minor variant of GZIP

**Reply by rsund** [reply]
The linked documentation says:
>All that differs in BGZF is that compressed size of each gzip block is limited to 2^16 bytes, and an extra ‘BC’ field in the gzip header records this size. Traditional decompression tools can ignore this, and unzip the file just like any other gzip file.

Could adding support be so simple that the mentioned non-standard header is allowed while [checking the gzip header](https://github.com/duckdb/duckdb/blob/a12329e657df33c4e47e62459e42760f1fe49610/src/common/gzip_file_system.cpp#L260)?

**Comment by Ed2uiz** [comment]
Commenting to also request bgzf support in duckdb (especially with the R API). bgzf compression is also used in BAM files which are used widely in the bioinformatics field for studying RNA sequencing data. R is the predominant language for most bioinformatics packages. With bgzf support in R, we could (hopefully) go straight from BAM to duckdb to biological insights in record speed.

**Comment by rsund** [comment]
PR for basic implementation in #6817.
